lightningd/dual_open_control.c

```diff
@@ -1568,8 +1597,118 @@ static void handle_peer_tx_sigs_msg(struct subd *dualopend,
 	wallet_channel_save(ld->wallet, channel);
 
 	/* Send notification with peer's signed PSBT */
-	notify_openchannel_peer_sigs(ld, &channel->cid,
-				     channel->psbt);
+	notify_openchannel_peer_sigs(ld, &channel->cid, chan_psbt);
+}
+
+static void handle_validate_rbf(struct subd *dualopend,
+				const u8 *msg)
+{
+	struct wally_psbt *candidate_psbt;
+	struct channel_inflight *inflight;
+	struct channel *channel = dualopend->channel;
+	struct wally_psbt *set_psbt;
+	bool *inputs_present;
+	struct amount_sat candidate_fee, last_fee;
+
+	if (!fromwire_dualopend_rbf_validate(tmpctx, msg,
+					     &candidate_psbt)) {
+		channel_internal_error(channel,
+				       "Malformed dualopend_rbf_validate %s",
+				       tal_hex(tmpctx, msg));
+		return;
+	}
+
+	/* Grab the first ever funding tx as starting point */
+	inflight = list_top(&channel->inflights, struct channel_inflight, list);
+	set_psbt = clone_psbt(NULL, inflight->funding_psbt);
```

Comment:
Don't need this clone any more, I think?

Reply:
we need a copy of the original PSBT to run comparisons against; we re-use `inflight` to iterate below.

Reply:
I don't think so.  Start with the candidate, compare it to the inflights:

```diff
diff --git a/lightningd/dual_open_control.c b/lightningd/dual_open_control.c
index 89a17ff63..fc8b01e55 100644
--- a/lightningd/dual_open_control.c
+++ b/lightningd/dual_open_control.c
@@ -1606,7 +1606,6 @@ static void handle_validate_rbf(struct subd *dualopend,
 	struct wally_psbt *candidate_psbt;
 	struct channel_inflight *inflight;
 	struct channel *channel = dualopend->channel;
-	struct wally_psbt *set_psbt;
 	bool *inputs_present;
 	struct amount_sat candidate_fee, last_fee;
 
@@ -1618,11 +1617,7 @@ static void handle_validate_rbf(struct subd *dualopend,
 		return;
 	}
 
-	/* Grab the first ever funding tx as starting point */
-	inflight = list_top(&channel->inflights, struct channel_inflight, list);
-	set_psbt = clone_psbt(NULL, inflight->funding_psbt);
-
-	inputs_present = tal_arr(tmpctx, bool, set_psbt->num_inputs);
+	inputs_present = tal_arr(tmpctx, bool, candidate_psbt->num_inputs);
 	memset(inputs_present, true, tal_bytelen(inputs_present));
 
 	/* Iterate through all previous inflights, confirm
@@ -1630,9 +1625,9 @@ static void handle_validate_rbf(struct subd *dualopend,
 	 * this one. */
 	list_for_each(&channel->inflights, inflight, list) {
 		/* Remove every non-matching input from set */
-		for (size_t i = 0; i < set_psbt->num_inputs; i++) {
+		for (size_t i = 0; i < candidate_psbt->num_inputs; i++) {
 			struct wally_tx_input *input =
-				&set_psbt->tx->inputs[i];
+				&candidate_psbt->tx->inputs[i];
 			struct bitcoin_txid in_txid;
 
 			wally_tx_input_get_txid(input, &in_txid);
@@ -1643,17 +1638,6 @@ static void handle_validate_rbf(struct subd *dualopend,
 		}
 	}
 
-	/* Finally, let's check the candidate psbt */
-	for (size_t i = 0; i < set_psbt->num_inputs; i++) {
-		struct wally_tx_input *input = &set_psbt->tx->inputs[i];
-		struct bitcoin_txid in_txid;
-
-		wally_tx_input_get_txid(input, &in_txid);
-
-		if (!psbt_has_input(candidate_psbt, &in_txid, input->index))
-			inputs_present[i] = false;
-	}
-
 	/* Are there any inputs that were present on all txs? */
 	if (memeqzero(inputs_present, tal_bytelen(inputs_present))) {
 		char *errmsg;
@@ -1708,7 +1692,6 @@ static void handle_validate_rbf(struct subd *dualopend,
 
 send_msg:
 	subd_send_msg(channel->owner, take(msg));
-	tal_free(set_psbt);
 }
```